util/logging.c

```diff
@@ -167,3 +176,103 @@ int nvme_submit_passthru64(int fd, unsigned long ioctl_cmd,
 
 	return err;
 }
+
+static void nvme_show_req_admin(const struct nvme_mi_admin_req_hdr *hdr, size_t hdr_len,
+				const void *data, size_t data_len)
+{
+	struct nvme_passthru_cmd cmd = {
+		.opcode = hdr->opcode,
+		.flags = hdr->flags,
+		.nsid = le32_to_cpu(hdr->cdw1),
+		.cdw2 = le32_to_cpu(hdr->cdw2),
+		.cdw3 = le32_to_cpu(hdr->cdw3),
+		.addr = (uint64_t)(uintptr_t)data,
+		.data_len = data_len,
+		.cdw10 = le32_to_cpu(hdr->cdw10),
+		.cdw11 = le32_to_cpu(hdr->cdw11),
+		.cdw12 = le32_to_cpu(hdr->cdw12),
+		.cdw13 = le32_to_cpu(hdr->cdw13),
+		.cdw14 = le32_to_cpu(hdr->cdw14),
+		.cdw15 = le32_to_cpu(hdr->cdw15),
+	};
+
+	nvme_show_common(&cmd);
+}
+
+static void nvme_show_req(__u8 type, const struct nvme_mi_msg_hdr *hdr, size_t hdr_len,
+			  const void *data, size_t data_len)
+{
+	if (type != NVME_MI_MSGTYPE_NVME)
+		return;
+
+	switch (hdr->nmp >> 3 & 0xf) {
+	case NVME_MI_MT_CONTROL:
+		break;
+	case NVME_MI_MT_MI:
+		break;
+	case NVME_MI_MT_ADMIN:
+		nvme_show_req_admin((struct nvme_mi_admin_req_hdr *)hdr, hdr_len, data, data_len);
+		break;
+	case NVME_MI_MT_PCIE:
+		break;
+	case NVME_MI_MT_AE:
+		break;
+	default:
+		break;
+	}
+}
+
+void *nvme_mi_submit_entry(__u8 type, const struct nvme_mi_msg_hdr *hdr, size_t hdr_len,
+			   const void *data, size_t data_len)
+{
+	memset(&sb, 0, sizeof(sb));
+
+	if (log_level >= LOG_DEBUG) {
+		nvme_show_req(type, hdr, hdr_len, data, data_len);
+		gettimeofday(&sb.start, NULL);
+	}
+
+	return &sb;
+}
+
+static void nvme_show_resp_admin(const struct nvme_mi_admin_resp_hdr *hdr, size_t hdr_len,
+				 const void *data, size_t data_len)
+{
+	printf("result       : %08x\n", hdr->cdw0);
+	printf("err          : %d\n", hdr->status);
```

Comment:
I assume we don't want the data here?

Reply:
Currently nvme_submit_passthru() for PCIe does not print the data so at this moment not required as same but later possible to be added the data output also I think.